### PR TITLE
Optimize clampScalar() for Vector2 and Vector3

### PR DESCRIPTION
**Pull request by leodutra**
Optimum code, removing closure from: 
- Vector2.clampScalar()
- Vector3.clampScalar()

I :heart: three.js. I'm fanboy since 2011 of @mrdoob and @alteredq productions.
And now, also you @Mugen87

I've been dealing with business systems for 14 years, now finally I'll start diving in 3D
and games.

### CODE REVIEW

File: src/math/Vector2.js [hunk 1]
```diff
@@ -278,21 +278,14 @@ Object.assign( Vector2.prototype, {
 
 	},
 
-	clampScalar: function () {
+	clampScalar: function (minVal, maxVal) {
```
Comment (Mugen87) [inline src/math/Vector2.js]: Minor code style issue. It should be:
```js
clampScalar: function ( minVal, maxVa l) {
```

File: src/math/Vector2.js [hunk 2]
```diff
@@ -388,10 +381,7 @@ Object.assign( Vector2.prototype, {
 		// computes the angle in radians with respect to the positive x-axis
 
 		var angle = Math.atan2( this.y, this.x );
-
-		if ( angle < 0 ) angle += 2 * Math.PI;
-
-		return angle;
+		return angle < 0 ? angle + 2 * Math.PI : angle;
```
Comment (Mugen87) [inline src/math/Vector2.js]: Maybe like so:
```js
return ( angle < 0 ) ? ( angle + 2 * Math.PI ) : angle;
```
It's a bit easer to read.

File: src/math/Vector2.js (hunk 1 above)
Comment (Mugen87) [reply src/math/Vector2.js]: In any event, resolving the closure is a good idea in order to reduce the amount of scopes (which can have an impact on performance).

File: src/math/Vector2.js [hunk 3]
```diff
@@ -379,7 +372,13 @@ Object.assign( Vector2.prototype, {
 
 	normalize: function () {
 
-		return this.divideScalar( this.length() || 1 );
+		var l = this.length();
```
Comment (Mugen87) [inline src/math/Vector2.js]: TBH, the previous version looked much better. I would revert this change.

File: src/math/Vector2.js (hunk 2 above)
Comment (leodutra) [reply src/math/Vector2.js]: OK.

File: src/math/Vector2.js (hunk 1 above)
Comment (leodutra) [reply src/math/Vector2.js]: OK.

File: src/math/Vector2.js (hunk 3 above)
Comment (leodutra) [reply src/math/Vector2.js]: Better but not faster.
There's a considerable cost on calling function vs. using expanded calculations.
I'd expand even .length line, but I kept as is because it is harder to mentally link expanded calculation to a length (missing the explanatory context).

I think calculation on a Vector2 is a priority, right?

Comment (Mugen87) [reply src/math/Vector2.js]: You should proof with a benchmark how much faster this calculation is. The the benefit is just minimal, I would keep the original version.

Comment (leodutra) [reply src/math/Vector2.js]: Yes, I'll re-sort this.
Actually just found out the code is formatted using tabs with a size of 2, and not 2 spaces.
The download is taking a long, long, long time.

Would be nice if secondary resources got split into different org projects.

**Comment by WestLangley** [comment]
I do not have objections to revisiting this, but if you have not done so, please read #11253 (and the embedded links) so you are familiar with the many views on this topic.

Also, the code was written for clarity, primarily, with the idea that it could be optimized for performance later if a problem was _demonstrated_.

**Comment by leodutra** [comment]
Hi @WestLangley. Checked the link.
In fact, the zero checking is kept, but the optimization is based on the simple knowledge a function call is more expensive than local variable access.

I can produce benchmarks. How do you guys do it? JsPerf, local benchmark print or any other way?

**Comment by Mugen87** [comment]
JsPerf is good.

**Comment by leodutra** [comment]
> JsPerf is fine.

Full source or just the function?

**Comment by leodutra** [comment]
The only justified change is for `<Vector>clampScalar()`.
https://jsperf.com/diff-speed-threejs-10-04-2019/1

**Comment by Mugen87** [comment]
The performance of the new `normalize()` method gets worse on my system. `angle()` stays more or less the same. But the new `.clampScalar()` is faster (from 51,719,130 to 86,304,699 Ops/sec). Probably because of the missing closure. In this case I would undo the change to `normalize()` and keep the rest.

BTW: It would be better to make 3 perf test, one for each method

**Comment by leodutra** [comment]
Added JS Perf for Vector3 too.
https://jsperf.com/diff-speed-threejs-10-04-2019/4

**Review by Mugen87 — APPROVED** [review]

**Comment by WestLangley** [comment]
Hmm. Performance is worse on my mobile device... Can you make a test for mobile-only so we can check several?

**Comment by Mugen87** [comment]
@WestLangley I have tested on my Pixel 1 with latest Chrome FF and the performance of the new approach is up to three times better. What mobile device/browser are you using?

**Comment by WestLangley** [comment]
@Mugen87 Safari on iPhone 6. Granted it is a bit old.

**Comment by leodutra** [comment]
In general I see IE and Safari with awkward and poor internal handling on
random things, depending on the version.

I'd test on a latest version and consider any relative difference something
which may change, drastically, on a new engine version.

That's also the case for normalize() and angle() I have just discarded.
According to my notes, dating from 2015, they would have a considerable
benefit from the use of ternary to have one less step and a plain
calculation would be always better than a method call.

Looks like engine empowered methods, probably because of AoT predictions
for the newest ES features.

IMHO, I think Chrome and FF are the end users would use, when they want
some morehigh performance on WebGL.

**Comment by WestLangley** [comment]
>IMHO, I think Chrome and FF are the end users would use, when they want
some more high performance on WebGL.

Almost all iOS users use Safari.

Can you make a test for mobile-only?

**Comment by Mugen87** [comment]
> Performance is worse on my mobile device...

How much worse is the performance compared to the current version?

**Comment by WestLangley** [comment]
>Can you please mentioned in numbers how the performance looks like before and after the change

See the jsperf test.

@Mugen87 Those are good reasons. If you don't care about further mobile tests, that is fine with me. This is not a hot method.

**Comment by Mugen87** [comment]
@mrdoob I think this is good to go 🚀

**Comment by mrdoob** [comment]
Thanks!
